### PR TITLE
Add cycling sequence workchain

### PR DESCRIPTION
**Pull request by ramirezfranciscof**

### CODE REVIEW

**Comment by edan-bainglass** [comment]
Workflow looks good! Per our discussion, we'll settle on a standard for techniques/protocols following user inquiry, then address monitoring after we upgrade to aiida-core 2.x.

File: aiida_aurora/workflows/cycling_sequence.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        return error_message
+
+
+class CyclingSequenceWorkChain(WorkChain):
+    """This workflow represents a process containing a variable number of steps."""
+
+    @classmethod
+    def define(cls, spec):
+        """Define the process specification."""
+        # yapf: disable
```
Comment (edan-bainglass) [inline aiida_aurora/workflows/cycling_sequence.py:33]: Why disable yapf?

File: aiida_aurora/workflows/cycling_sequence.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            help="Battery sample to be used."
+        )
+        spec.input(
+            "tomato_code",
+            valid_type=orm.Code,
+            help="Tomato code to use."
+        )
+
+        spec.input_namespace(
+            "techniques",
```
Comment (edan-bainglass) [inline aiida_aurora/workflows/cycling_sequence.py:47]: Need to settle on a standard for techniques/protocols through user inquiry. Checking...

**Review by edan-bainglass — COMMENTED** [review]
Workflow looks good! Per our discussion, we'll settle on a standard for techniques/protocols following user inquiry, then address monitoring after we upgrade to aiida-core 2.x.

File: aiida_aurora/workflows/cycling_sequence.py (hunk 2 above)
Comment (edan-bainglass) [reply aiida_aurora/workflows/cycling_sequence.py:47]: Received user feedback. This should be `protocols`, which is a collection of individual protocols, each being a sequence of `techniques`.

**Review by edan-bainglass — APPROVED** [review]
LGTM! However, some of your changes will no longer be necessary due to updates on my end:

- `Rename cycling sequence test` - I updated the CI in #27 to only run `pytest` on the _tests_ directory
- `Fix errors on CI due to pydantic` - I pinned pydantic's version in #26

Maybe check/approve #26 and #27 and rebase this on top?

File: aiida_aurora/workflows/cycling_sequence.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            self.report(f'Cycling substep <pk={pkid}> failed with exit status {stat}')
+            return self.exit_codes.ERROR_IN_CYCLING_STEP
+
+    def run_cycling_step(self):
+        """Run the next cycling step."""
+        current_keyname = self.worksteps_keynames.pop(0)
+        inputs = {
+            'code': self.inputs.tomato_code,
+            'battery_sample': self.inputs.battery_sample,
+            'technique': self.inputs.protocols[current_keyname],
```
Comment (edan-bainglass) [inline aiida_aurora/workflows/cycling_sequence.py:110]: `'technique'` should be `'protocol'`. As discussed, changes need to be made in `cycler.py` to uniform the language. As a reminder, `protocols` are a set of ``protocol` instances, e.g. "formation", "cycling", each made up of a set of `technique` instances, e.g. "OCV", "CC", "LOOP", etc.

File: examples/cycling_sequence/test_cycling_sequence.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+# ACTUAL SUBMISSION
+
+WorkflowClass = WorkflowFactory('aurora.cycling_sequence')
+workflow_builder = WorkflowClass.get_builder()
+workflow_inputs = generate_test_inputs()
+
+workflow_builder.tomato_code = orm.load_code('ketchup-0.2rc2@localhost-tomato')
+
+workflow_builder.battery_sample = workflow_inputs['sample']
+workflow_builder.techniques = {
```
Comment (edan-bainglass) [inline examples/cycling_sequence/test_cycling_sequence.py:99]: `workflow_builder.protocols`

**Review by edan-bainglass — CHANGES_REQUESTED** [review]
There is still a confusion between protocol and technique concepts. It seems to me that both singular and plural cases of protocol rely on `ElectroChemSequence` as a schema. But this surely is incorrect, right? The plural is of course a set of the singular, but the singular is a set of techniques, so different. Need to discuss/clarify.

File: aiida_aurora/workflows/cycling_sequence.py (hunk 3 above)
Comment (ramirezfranciscof) [inline aiida_aurora/workflows/cycling_sequence.py:110]:
```suggestion
            'protocol': self.inputs.protocols[current_keyname],
```

File: aiida_aurora/workflows/cycling_sequence.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        )
+
+        spec.exit_code(
+            401,
+            'ERROR_IN_CYCLING_STEP',
+            message='One of the steps of CyclingSequenceWorkChain failed'
+        )
+
+    def setup_workload(self):
+        """Takes the inputs and wraps them together."""
```
Comment (ramirezfranciscof) [inline aiida_aurora/workflows/cycling_sequence.py:87]:
```suggestion
        """Take the inputs and wrap them together."""
```

File: examples/cycling_sequence/test_cycling_sequence.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+from aiida.plugins import DataFactory, WorkflowFactory
+
+CyclingSpecsData = DataFactory('aurora.cyclingspecs')
+BatterySampleData = DataFactory('aurora.batterysample')
+TomatoSettingsData = DataFactory('aurora.tomatosettings')
+
+
+@calcfunction
+def generate_test_inputs():
+    """Generated the inputs."""
```
Comment (ramirezfranciscof) [inline examples/cycling_sequence/test_cycling_sequence.py:17]:
```suggestion
    """Generate the inputs."""
```

File: examples/cycling_sequence/test_cycling_sequence.py (hunk 4 above)
Comment (ramirezfranciscof) [inline examples/cycling_sequence/test_cycling_sequence.py:99]:
```suggestion
workflow_builder.protocols = {
```

**Review by ramirezfranciscof — COMMENTED** [review]
Changes to apply

**Comment by edan-bainglass** [comment]
I still don't understand. You did all of this in response to my request for changes? It was not already there?

**Comment by edan-bainglass** [comment]
Review recap (detailed and accurate 🤞)
======================

1. Approved as is (too hasty - found this out once I started modifying the workflow for monitoring)
2. Noticed you addressing items that have been taken care of in the AiiDA 2.x upgrade and were no longer necessary (fixed 👍)
3. Noticed that `technique` was not changed to `protocol` as discussed (fixed 👍)
4. Found `technique` references in `cycler.py` (yet to be fixed)

This brings us to the present. Now, regarding my comment about too many changes in this latest round, that's my bad. I was looking at ALL PR commits, rather than just the commit details of `FIX: Corrections from PR review`. That looks good to me, but it still does not address review 4. At this point, do you mind if I address the rest? I need to finalize this thing by end of week, and really, I only have today (packing and moving to new office Th/F).

Let me know 🙏

**Comment by ramirezfranciscof** [comment]
> The commit looks good to me, but it still does not address review 4. At this point, do you mind if I address the rest? I need to finalize this thing by end of week, and really, I only have today (packing and moving to new office Th/F).

Sure, no problem you can go ahead. But I think I applied all the changes from the previous reviews (except maybe the yapf ignore statements that I think just make the specs declaration harder to read). Which one is review 4, [this one](https://github.com/epfl-theos/aiida-aurora/pull/19#pullrequestreview-1600521958)? I answered the questions but I'm not sure what else is there to address...

**Comment by edan-bainglass** [comment]
In my previous message, 1-3 are done! 4 was not yet addressed, i.e. you changed the reference to the calculation's `technique` in the workflow, but not in the actual calcjob.

In any case, I'll rebase and address the rest, including my changes. You can review all that in a bit.

**Comment by edan-bainglass** [comment]
By the way, how would I go about proceeding? Do I push my changes to your fork, so that it is reflected in the PR? Or is there another way?

**Pull request merged**